### PR TITLE
Fix bug with -ffp-contract=fast-honor-pragmas

### LINKED ISSUE #104830: clang asserts with ffp-contract=fast-honor-pragmas
When I use "#pragma STDC FP_CONTRACT DEFAULT" and compile with "-ffp-contract=fast-honor-pragmas", clang asserts.

```
float f(float a, float b, float c) {
    #pragma STDC FP_CONTRACT DEFAULT
    return a * b + c;
}
```

See:

https://godbolt.org/z/EqE6Kn6Wf

### PR DESCRIPTION
**Pull request by andykaylor**
This fixes a problem which caused clang to assert in the Sema pragma
handling if it encountered "#pragma STDC FP_CONTRACT DEFAULT" when
compiling with the -ffp-contract=fast-honor-pragmas option.

This fixes https://github.com/llvm/llvm-project/issues/104830

### CODE REVIEW

File: clang/test/CodeGen/ffp-contract-fast-honor-pramga-option.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  a[0] -= b * c;
+}
+
+float fp_contract_5(float a, float b, float c) {
+  // CHECK-LABEL: fp_contract_5fff(
+  // CHECK: fmul contract float
+  // CHECK: fadd contract float
+  float t = a * b;
+  return t + c;
+}
```
Comment (AaronBallman) [inline clang/test/CodeGen/ffp-contract-fast-honor-pramga-option.cpp:37]: Please add a newline to the end of the file.

File: clang/test/CodeGen/ffp-contract-fhp-pragma-override.cpp [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  return a * b + c;
+}
+
+float fp_contract_global_default_2(float a, float b, float c) {
+  // CHECK-LABEL: fp_contract_global_default_2fff(
+  // CHECK: fmul contract float
+  // CHECK: fadd contract float
+  float t = a * b;
+  return t + c;
+}
```
Comment (AaronBallman) [inline clang/test/CodeGen/ffp-contract-fhp-pragma-override.cpp:151]: And a newline here as well.

**Review by AaronBallman — APPROVED** [review]
Changes LGTM modulo nits, but please also add a release note to `clang/docs/ReleaseNotes.rst` so users know about the fix.

**Review by zahiraam — APPROVED** [review]
LGTM. Thanks.
I had to double check, but it looks like the LIT test `ffp-contract-fast-honor-pragma-option.cpp` has the same behavior with and without this patch. Right? Aren't there already tests that check the output of the option?

**Comment by andykaylor** [comment]
> LGTM. Thanks. I had to double check, but it looks like the LIT test `ffp-contract-fast-honor-pragma-option.cpp` has the same behavior with and without this patch. Right? Aren't there already tests that check the output of the option?

Yes, the `ffp-contract-fast-honor-pragma-option.cpp` is testing existing behavior. The existing tests didn't cover this.

**Review by jcranmer-intel — APPROVED** [review]

**Pull request merged**